File: merge sort/mergesort.py
# Christian Roccanova
# CS 325 Fall 2018
# Mergesort - sorts data from file by splitting it into subarrays and recombining them

#merges left and right sub arrays
def merge(data, left, right):
     
       #array indices
       i = 0
       j = 0
       k = 0

       sizeL = len(left)
       sizeR = len(right)

       #merge sub arrays 
       while i < sizeL and j < sizeR:
           if left[i] < right[j]:
               data[k] = left[i]
               i += 1
           else:
               data[k] = right[j]
               j += 1
           k += 1

       #copy remaining elements of left sub array
       while i < sizeL:
           data[k] = left[i]
           i += 1
           k += 1
       #copy remaining elements of right sub array
       while j < sizeR:
           data[k] = right[j]
           j += 1
           k += 1

#recursively sorts array by splitting it into sub arrays, sorting those and then re-combining them
def mergeSort(data):

   dataSize = len(data)

   #if only one element, its already sorted
   if dataSize > 1:
       
       #split passed data into left and right halves
       mid = int(dataSize/2) #get position of middle and force it to be an int
       left = data[:mid] #array of all elements from position 0 to mid
       right = data[mid:] #array of all elements from mid to last element


       #recursive calls on each split
       mergeSort(left)
       mergeSort(right)

       #re-combines sub arrays
       merge(data, left, right)

# Main 
#get data from file
dataFile = open("data.txt", "r")
lines = dataFile.readlines()
dataFile.close()
lineCount = len(lines)

#separate lines
for i in range(0, lineCount):
    sortData = []
    #separate line elements
    currLine = lines[i].split()    
    dataLength = int(currLine[0])
    
    #append line elements to data array
    for j in range(1, dataLength+1):
        sortData.append(int(currLine[j]))

    #call insertion sort
    mergeSort(sortData)

    #write data to output file
    mergeOut = open("merge.out", "a")
    for k in range (0, dataLength):
        mergeOut.write(str(sortData[k]))
        mergeOut.write(" ")
    mergeOut.write("\n")
    mergeOut.close()

File: merge sort/mergesort.cpp
/******************************************************
Program Name: mergesort
Author: Christian Roccanova
Date: 4/6/2018
Description: reads data from file into an array and then sorts via merge sort.  Sorted data is then output to merge.out
******************************************************/

#include<iostream>
#include <fstream>
#include <string>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::ifstream;
using std::ofstream;

//compares values and stores them in a temporary array in sorted order to be copied over the original array at the end
void merge(int *arr, int low, int mid, int high)
{
	int indexA = low;
	int indexB = mid + 1;
	int indexC = low;
	int temp[100000];	//temporary array holds sorted values

	//compares elements and adds the lower to the temporary array
	while (indexA <= mid && indexB <= high)
	{
		if (arr[indexA] < arr[indexB])
		{
			temp[indexC] = arr[indexA];
			indexA++;
			indexC++;
					
		}
		else
		{
			temp[indexC] = arr[indexB];
			indexB++;
			indexC++;
		}
	}

	while (indexA <= mid)
	{
		temp[indexC] = arr[indexA];
		indexA++;
		indexC++;
	}

	while (indexB <= mid)
	{
		temp[indexC] = arr[indexB];
		indexB++;
		indexC++;
	}

	//copies sorted temporary array into original array
	for (int i = low; i < indexC; i++)
	{
		arr[i] = temp[i];
	}
}

//sorts array by breaking down the input array and then calling merge() to put it back together in sorted order
void mergeSort(int *arr, int low, int high)
{
	//if array size is greater than 0, recursively calls mergeSort
	if (low < high)
	{
		//sets midpoint
		int mid = low + (high - low) / 2;

		mergeSort(arr, low, mid);
		mergeSort(arr, mid + 1, high);
		merge(arr, low, mid, high);

	}
}

int main()
{
	int rowLen;
	int rowTotal = 0;
	int rowCount = 0;
	int numArr[100000];
	string rowString;
	ifstream dataFile;
	ofstream outFile("merge.out");

	
	dataFile.open("data.txt");
	//counts number of rows in data file
	while (!dataFile.eof()) {
		getline(dataFile, rowString);
		rowTotal++;
	}
	
	dataFile.close();

	dataFile.open("data.txt");

	//cycles through each row individually
	for (int i = 0; i < rowTotal; i++)
	{

		dataFile >> rowLen;

		//fills initial array
		for (int j = 0; j < rowLen; j++)
		{
			dataFile >> numArr[j];
		}

		mergeSort(numArr, 0, rowLen-1);

		//outputs sorted array to file
		for (int j = 0; j < rowLen; j++)
		{
			outFile << numArr[j] << " ";
		}


		outFile << endl;


	}
	outFile.close();
	dataFile.close();

	return 0;
}